File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    private HashMap<String, String> info = new HashMap<String, String>();
    private List<String[]> prints = new ArrayList<String[]>();
    
    public String[] parse(String str) {
        return str.split(" ");
    }
    
    public void updateInfo(String[] strs) {
        String command = strs[0], uid = strs[1];
        
        if (command.equals("Enter")) {
            String[] print = {uid, "님이 들어왔습니다."};
            prints.add(print);
            info.put(uid, strs[2]);
        }
        
        else if (command.equals("Leave")) {
            String[] print = {uid, "님이 나갔습니다."};
            prints.add(print);
        }
            
        else
            info.put(uid, strs[2]);
    }
    
    public String[] solution(String[] record) {
        
        for (String r : record) {
            String[] str = parse(r);
            updateInfo(str);
        }
        
        String[] answer = new String[prints.size()];
        
        int idx = 0;
        for (String[] print : prints) {
            String p = info.get(print[0]) + print[1];
            answer[idx++] = p;
        }
        return answer;
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;
vector<string> parse(string str) {
    vector<string> strs;
    
    int idx = 0, len = 0;
    for (char c : str) {
        if (c == ' ') {
            strs.push_back(str.substr(idx, len));
            cout << str.substr(idx, len) << endl;
            idx += len;
            len = 0;
        }
        else
            len++;
        // cout << c << endl;
    }
    strs.push_back(str.substr(idx, len));
    return strs;
}

vector<string> solution(vector<string> record) {
    vector<string> answer;
    map<string, string> info;
    
    for (string r : record) {
        parse(r);    
    }
    
    return answer;
}